### ISSUE TITLE
WASM interoperability

### ISSUE THREAD

**Issue by jeromesimeon** [accordproject/ergo#727]
**Is your feature request related to a problem? Please describe.**

Would anyone like to call WASM from Ergo? WASM 

**Describe the solution you'd like**

I don't know yet what I'd like!

**Additional context**

WASM specification: https://webassembly.github.io/spec/core/

Reference implementation in OCaml: https://github.com/WebAssembly/spec

WASM/JS interface: https://webassembly.github.io/spec/js-api/index.html

Interesting bits about WASM/JS interop performances: https://hacks.mozilla.org/2018/10/calls-between-javascript-and-webassembly-are-finally-fast-🎉/

**Comment by jeromesimeon** [comment]
Things we might need to get this working:
- a cool application
- need an Ergo syntax or API (for calling a foreign function)
- a mapping between the two type systems See notably https://webassembly.github.io/spec/core/syntax/values.html

**Comment by AnalystAbhishek** [comment]
I would to work on this project.

**Comment by irmerk** [comment]
Great @AnalystAbhishek! If you're interested in the GSoC project around WASM, be sure to look into the resources we have pinned in our slack channel.

**Comment by AnalystAbhishek** [comment]
Yeah I'm already going through the post shared by Dan and Jerome on slack channel.

**Comment by AnalystAbhishek** [comment]
I'have finished my first task of understanding the whole documentation of Accord Project and WASM. I have also set up the enviroment to start work on it.

**Comment by jeromesimeon** [comment]
Some notes from today's discussion on the WASM backend.

## Plan

1. ImpWasm = Imp over EJson+SumTypes (i.e., do not encode left/right as objects) --- Jerome

2. Mini compiler: take imp_ejson AST (just constant + a few operators) and compile to WASM AST + (static) link runtime + run -- Patrik

3. Grow the runtime supported (& decide which runtime operators can be eliminated by the compiler) -- Later

4. Grow the subset of imp_ejson AST being compiled
   1. if-then-else
   2. statements: for loops
   3. blocks?? probably need to eliminate blocks because of scoping

## Blocks & scoping --- Jerome

WASM doesn't support block-level scoping (not 'let' only 'var')

Imp supports block-level scoping. e.g.,:
```
{ let x = 1;
  let y = 3;
  { let x = 2;
    x + y } }
```

Maybe removed by renaming/blocks merging:
```
{ let x1 = 1 ;
  let y = 3;
  let x2 = 2;
  x2 + y }
```

## Runtime operators

Examples of runtime operators

```
concat: [a:1,b:2] + [c:3]     ==> [a:1,b:2,c:3]
concat: [a:1,b:2] + [c:3,b:4] ==> [a:1,c:3,b:4]

merge: [a:1,b:2] + [c:3]     ==> some([a:1,b:2,c:3])
merge: [a:1,b:2] + [c:3,b:2] ==> some([a:1,b:2,c:3])
merge: [a:1,b:2] + [c:3,b:4] ==> none

nth: [1,2,3] + 0 ==> some(1)
nth: [1,2,3] + 3 ==> none
```

**Comment by jeromesimeon** [comment]
Todo list:
I think a possible todo list:
1. [ ] Get the release-1.0 branch of Ergo to build on the wasm branch of Q*cert (requires per-module extraction)
2. [ ] Add a -target wasm option to the Ergo compiler, hook it up to a corresponding ErgoImptoWasm translation backed up by your compiler
3. [ ] Add a -target imp_ejson option to the Ergo compiler? Along with a proper pretty printer consistent with the new imp_ejson syntax we saw in the demo
4. [ ] Figure out runtime/testing, missing runtime operators for helloworld etc.
Most likely: work on compilation-time elimination of brand operations (in Q*cert, removing them from Imp runtime)
5. [ ] [Not Wasm Specific] Still need to update release-1.0 of Ergo to use the latest development version of Concerto (pre Concerto 1.0)
6. [ ] [Not Wasm Specific] Merge release-1.0 branch of Ergo into master
Merge wasm branch of Q*cert into master ; publish Q*cert ; publish Concerto ; publish Ergo
7. [ ] Tada!
